packages/remirror__extension-annotation/src/annotation-plugin.ts

```diff
@@ -24,7 +24,92 @@ export class AnnotationState<Type extends Annotation = Annotation> {
    */
   decorationSet = DecorationSet.empty;
 
-  constructor(private readonly getStyle: GetStyle<Type>) {}
+  constructor(
+    private readonly getStyle: GetStyle<Type>,
+    private readonly map: MapLike<string, OmitText<Type>>,
+    private readonly transformPosition: (pos: number) => number,
+    private readonly transformPositionBeforeRender: (pos: number) => number | null,
+  ) {}
+
+  addAnnotation(addAction: AddAnnotationAction<Type>): void {
+    const { id } = addAction.annotationData;
+    this.map.set(id, {
+      ...addAction.annotationData,
+      from: this.transformPosition(addAction.from),
+      to: this.transformPosition(addAction.to),
+    } as OmitText<Type>);
+  }
+
+  updateAnnotation(updateAction: UpdateAnnotationAction<Type>): void {
+    assert(this.map.has(updateAction.annotationId));
+
+    this.map.set(updateAction.annotationId, {
+      ...this.map.get(updateAction.annotationId),
+      ...updateAction.annotationData,
+    } as OmitText<Type>);
+  }
+
+  removeAnnotations(removeAction: RemoveAnnotationsAction): void {
+    removeAction.annotationIds.forEach((id) => {
+      this.map.delete(id);
+    });
+  }
+
+  setAnnotations(setAction: SetAnnotationsAction<Type>): void {
+    // YJS maps don't support clear
+    this.map.clear?.();
+    // eslint-disable-next-line prefer-arrow-callback
```

Comment:
Out of interest: Why? You seem to be using arrow callbacks in other places?

Reply:
Ah this is because the [third argument is `this`](https://github.com/yjs/yjs/blob/main/src/types/YMap.js#L175), so using an arrow function would bind the function's scope to the plugin's scope.

YJS's `forEach` does not provide a `thisArg` argument [like a native Map does](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/forEach#syntax), so I can't

```ts
this.map.forEach((_, id, map) => {
  map.delete(id);
}, map);
```

Does that make sense? I suppose I could do this, but seems icky to me.

```ts
this.map.forEach((_, id, map) => {
  this.map.delete(id);
});
```

Reply:
Ah, line 84... good spot. I'll look into it. Obviously I'll resolve the consistency as you know where I live 😨 😆